server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -129,6 +150,147 @@ public CompletionStage<Result> review(
+                v -> checkApplicantAuthorization(profileUtils, request, applicantId),
+                httpExecutionContext.current())
+            .toCompletableFuture();
+
+    CompletableFuture<ReadOnlyApplicantProgramService> applicantProgramServiceCompletableFuture =
+        applicantStage
+            .thenComposeAsync(
+                v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
+                httpExecutionContext.current())
+            .toCompletableFuture();
+
+    return CompletableFuture.allOf(
+            applicantAuthCompletableFuture, applicantProgramServiceCompletableFuture)
+        .thenApplyAsync(
+            (v) -> {
+              ReadOnlyApplicantProgramService roApplicantProgramService =
+                  applicantProgramServiceCompletableFuture.join();
+              Optional<Block> block = roApplicantProgramService.getBlock(blockId);
+              Optional<String> maybeAddressJson = request.session().get(ADDRESS_JSON_SESSION_KEY);
+
+              if (block.isEmpty()) {
+                return notFound();
+              }
+
+              if (maybeAddressJson.isEmpty()) {
+                throw new RuntimeException("Address JSON missing");
+              }
+
+              try {
+                Block thisBlockUpdated = block.get();
+                ObjectMapper mapper = new ObjectMapper();
+                JsonNode rootNode = mapper.readTree(maybeAddressJson.get());
+
+                ImmutableList.Builder<AddressSuggestion> suggestionBuilder =
+                    ImmutableList.builder();
+                for (JsonNode node : rootNode) {
+                  suggestionBuilder.add(AddressSuggestion.builder().fromJson(node));
+                }
+
+                ImmutableList<ApplicantQuestion> questions = thisBlockUpdated.getQuestions();
+
+                Optional<ApplicantQuestion> addressQuestionMaybe =
+                    questions.stream()
+                        .filter(
+                            m ->
+                                m.getQuestionDefinition().isAddress()
+                                    && m.isAddressCorrectionEnabled())
+                        .findFirst();
+
+                if (addressQuestionMaybe.isPresent()) {
+                  AddressQuestion addressQuestion =
+                      addressQuestionMaybe.get().createAddressQuestion();
+                  DynamicForm form = formFactory.form().bindFromRequest(request);
+                  String selectedAddress =
+                      form.get(AddressCorrectionBlockView.SELECTED_ADDRESS_NAME);
+
+                  Optional<AddressSuggestion> suggestionMaybe =
+                      suggestionBuilder.build().stream()
+                          .filter(x -> x.getSingleLineAddress().equals(selectedAddress))
+                          .findFirst();
+
+                  Map<String, String> map = new HashMap<>(Map.of());
+
+                  if (suggestionMaybe.isPresent()) {
```

Comment:
Not necessarily. If they could choose the option to keep the address as entered.

Reply:
If a user chooses to select the address as they entered it, then it can't be used for verification unless it takes the location from the top match.

It may be reasonable to not allow for selecting address as entered, if the block has an address service area validation predicate.

Reply:
View is set to show/hide the option to keep the address the user entered. Currently hard-coded to always show. Need to switch that to check if eligibility is required, but I'm not sure where to call that.

Reply:
That hard-coded comment is no longer valid and has been fixed.